src/library/scala/Enumeration.scala

```diff
@@ -294,6 +294,7 @@ abstract class Enumeration (initial: Int) extends Serializable {
     /** Creates a bit mask for the zero-adjusted ids in this set as a
      *  new array of longs */
     def toBitMask: Array[Long] = nnIds.toBitMask
+    private[Enumeration] lazy val byName: Map[String, Value] = iterator.map( v => v.toString -> v).toMap
```

Comment:
While the documentation warns against modifying an enumeration after construction, it explains that this is for thread safety. It looks like it is safe for single-threaded use:

>  Values SHOULD NOT be added to an enumeration after its construction;
  doing so makes the enumeration thread-unsafe. If values are added to an
  enumeration from multiple threads (in a non-synchronized fashion) after
  construction, the behavior of the enumeration is undefined.

Adding any memoized values like `byName` breaks this assumption.

Reply:
Hmm.  I thought that assumption can't've been in place given the `vset` caching, but I see now how `vsetDefined` is reset when a new `Val` is instantiated.   I guess we could recover this using the same pattern.

Reply:
Actually, looks like that assumption is already pretty broken even single-threaded:

```
Welcome to Scala 2.13.1 (OpenJDK 64-Bit Server VM, Java 11.0.6).
Type in expressions for evaluation. Or try :help.

scala> object Foo extends Enumeration {
     |   class Val extends super.Val
     |   val Bar, Baz = new Val
     | }
defined object Foo

scala> new Foo.Val
res0: Foo.Val = <Invalid enum: no field for #2>

scala> Foo.values
res1: Foo.ValueSet = Foo.ValueSet(Bar, Baz, <Invalid enum: no field for #2>)
```

Reply:
@dwijnand reminds me that the first day I learned some Scala I lodged https://github.com/scala/bug/issues/3719

There is a mechanism, or a couple of mechanisms:
```
scala> object X extends Enumeration { nextName = Iterator("a","b") ; class V extends Val ; val A = new V }
defined object X

scala> X.A
res0: X.V = a

scala> val B = new X.V
B: X.V = b

scala> val C = new X.V
C: X.V = <Invalid enum: no field for #2>
```
Once I suggested a name map for a java enum and now it's like every enum has to have one. I'm not sure scala enumeration is supposed to be used angrily enough for it to matter. Not sure if the onus is on the client to memoize.